include/envoy/config/config_provider_manager.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/config/config_provider.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * A ConfigProvider manager which instantiates static and dynamic (xDS) providers.
```

Comment:
In SDS, there are 3 kind providers which is static/inline/dynamic, in RDS we called it static is actually inline (inlined in HCM in LDS), we should be consistent in that wording.

Reply:
I spent some time looking at the SecretManagerImpl and ContextConfigImpl, and based on my understanding so far, the difference between static and inline providers is that the former are added to a map and shared via findStatic*Provider() while the latter are not.

From an implementation standpoint, they share the config provider impls (TlsCertificateConfigProviderImpl and CertificateValidationContextConfigProviderImpl) and obtain the config proto via the bootstrap config (for static) or listener config (for inline).

So, I agree it makes sense to have consistent wording, although I don't think the naming conventions applied to the concrete config providers should apply to these base classes. 

The reason the StaticConfigProviderImpl is prefixed with "Static" is that the config proto and impl are immutable after instantiation.

I propose either keeping the StaticConfigProviderImpl name for the base class and renaming the route config provider that will eventually subclass it to InlineStaticRouteConfigProvider, or changing the name of the base class to ImmutableConfigProviderImplBase to differentiate with the naming conventions already established for the concrete types.

Reply:
From config perspective, the difference is where the proto is coming from:
- static: from static_resources in [bootstrap.proto](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/bootstrap/v2/bootstrap.proto#L51)
- inline: inlined in cluster/listener config, regardless of the cluster/listener being static/dynamic.
- dynamic: delivered from xDS API

Secret is the only resource that have all three type above for several reasons. We don't have static route because there wasn't such demand. Listener/Cluster cannot be inline.

> I propose either keeping the StaticConfigProviderImpl name for the base class and renaming the route config provider that will eventually subclass it to InlineStaticRouteConfigProvider, or changing the name of the base class to ImmutableConfigProviderImplBase to differentiate with the naming conventions already established for the concrete types.

I prefer latter, "static" sounds like the config exists once they are instantiated and won't disappear. This is not true for "static route" here but is true for "static listener/cluster", so I think avoid "static" here would be less confusing.

Reply:
Ok. Unless there are any objections from the other reviewers, I'm going to rename both Static and Dynamic config providers to {Immutable,Mutable} respectively, so that the new wording is consistent.

Reply:
Renamed {Static,Dynamic}ConfigProviderImplBase to {Immutable,Mutable}. 

Also, formalized the notion of config types assigned to instances of ConfigProviders by introducing a ConfigProviderInstanceType.

PTAL. Thanks!

Reply:
Is there any way to simplify this from 3 types to 2 while we're doing this overhaul? @lizan can you elaborate on why we need the inline type here?